src/lib.rs

```diff
@@ -390,58 +409,65 @@ pub trait Trait: system::Trait + Sized {
-        // - `Foo: Option<u32>`:
-        //   - `Foo::put(1); Foo::get()` returns `Some(1)`;
-        //   - `Foo::kill(); Foo::get()` returns `None`.
-        // - `Foo: u32`:
-        //   - `Foo::put(1); Foo::get()` returns `1`;
-        //   - `Foo::kill(); Foo::get()` returns `0` (u32::default()).
-        // e.g. Foo: u32;
-        // e.g. pub Bar get(bar): map T::AccountId => Vec<(BalanceOf<T>, u64)>;
-        //
-        // For basic value items, you'll get a type which implements
-        // `support::StorageValue`. For map items, you'll get a type which
-        // implements `support::StorageMap`.
-        //
-        // If they have a getter (`get(getter_name)`), then your module will come
-        // equipped with `fn getter_name() -> Type` for basic value items or
-        // `fn getter_name(key: KeyType) -> ValueType` for map items.
-        Dummy get(dummy) config(): Option<BalanceOf<T>>;
+    trait Store for Module<T: Trait> as Forum {
+
+        /// Map category identifier to corresponding category.
+        pub CategoryById get(category_by_id) config(): map CategoryId => Category<T::BlockNumber, T::Moment, T::AccountId>;
+
+        /// Category identifier value to be used for the next Category created.
+        pub NextCategoryId get(next_category_id) config(): CategoryId;
+
+        /// Map thread identifier to corresponding thread.
+        pub ThreadById get(thread_by_id): map ThreadId => Thread<T::BlockNumber, T::Moment, T::AccountId>;
+
+        /// Thread identifier value to be used for next Thread in threadById.
+        pub NextThreadId get(next_thread_id) config(): ThreadId;
 
-        // A map that has enumerable entries.
-        Bar get(bar) config(): linked_map T::AccountId => BalanceOf<T>;
+        /// Map post identifier to corresponding post.
+        pub PostById get(post_by_id): map PostId => Post<T::BlockNumber, T::Moment, T::AccountId>;
 
-        // this one uses the default, we'll demonstrate the usage of 'mutate' API.
-        Foo get(foo) config(): BalanceOf<T>;
+        /// Post identifier value to be used for for next post created.
+        pub NextPostId get(next_post_id) config(): PostId;
 
+        /// Account of forum sudo.
+        pub ForumSudo get(forum_sudo) config(): T::AccountId;
 
-        // ==== Constraints ====
+        // === Add constraints here ===
 
     }
 }
 
 decl_event!(
-    /// Events are a simple means of reporting specific conditions and
-    /// circumstances that have happened that users, Dapps and/or chain explorers would find
-    /// interesting and otherwise difficult to detect.
     pub enum Event<T>
     where
-        B = BalanceOf<T>,
+        <T as system::Trait>::AccountId,
     {
-        // Just a normal `enum`, here's a dummy event to ensure it compiles.
-        /// Dummy event, just here so there's a generic type that's used.
-        Dummy(B),
+        /// A category was introduced
+        CategoryCreated(CategoryId),
```

Comment:
Did you consider to add extrinsic's `origin` as the first param of every event?
If you think of this as: `who` made `what`? Then it makes sense to add this `who` to an event.
```suggestion
        CategoryCreated(AccountId, CategoryId),
```

Reply:
1) unless I am mistaken `origin` is not a type, is an instance variable. The `who`variable often named in Substrate code is indeed of type `T::AccountId` as this.

2) I don't see a reason to add it to every event, what we want is to give the event recipient a way to learn more about the event by knowing a key identifier value. Picking the origin account is arbitrary, we can add many other things if we start down that road. The core point of an event is to allow anyone to learn that _something_ took place, not everything about what this thing was.

Reply:
Yes, your answers make sense. But as an example, when `who` could be useful: a client could be interested to subscribe/filter events produced by actions of a particular account.

Reply:
I am not sure I agree that there is an important use case of clients subscribing to events of specific other accounts that are not themselves, but either way, this sort of filtering/subscription has to be done by indexing nodes, not full nodes directly. These nodes will know the origin and prove it succinctly. 

Client needs should only impact runtime to give proper security guarantees.